### PR TITLE
Adding functionality for create cluster (CloudStack)

### PR DESCRIPTION
**Pull request by maxdrib**
*Issue #, if available:*

*Description of changes:*
This PR adds support for cluster creation using the Cloudstack provider. A few notes

1. This code is not yet integrated with CloudMonkey, which is still in PR, so the preflight check is a no-op
2. The default bundle manifest does not yet have a cloudstack object so all my testing was done with a bundle override which included the latest CAPC artifacts (hosted locally with nginx)
3. Unit tests have not been fully implemented yet and may be broken
4. Eks-a controller changes make it so the creation only succeeds with a well timed tilt execution
5. I haven't done any linting yet
6. This code doesn't support WorkerNodeGroups as it stands

Much of this code was originally written by @wanyufe so kudos to him

*Testing (if applicable):*
Cluster creation succeeded when using tilt and bundle overrides with shared network using 1 CP, 1 etcd, and 1 worker node.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/cloudstack/cloudstack.go [hunk 1]
```diff
@@ -0,0 +1,1159 @@
...
+	"text/template"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/aws/eks-anywhere/pkg/crypto"
+
+	etcdv1beta1 "github.com/mrajashree/etcdadm-controller/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	kubeadmnv1beta1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
```
Comment (vivek-koppuru) [inline pkg/providers/cloudstack/cloudstack.go:26]:
```suggestion
	kubeadmv1beta1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
```

File: pkg/executables/kubectl.go [hunk 2]
```diff
@@ -37,6 +37,8 @@ var (
 	eksaClusterResourceType           = fmt.Sprintf("clusters.%s", v1alpha1.GroupVersion.Group)
 	eksaVSphereDatacenterResourceType = fmt.Sprintf("vspheredatacenterconfigs.%s", v1alpha1.GroupVersion.Group)
 	eksaVSphereMachineResourceType    = fmt.Sprintf("vspheremachineconfigs.%s", v1alpha1.GroupVersion.Group)
+	eksaCloudStackDatacenterResourceType = fmt.Sprintf("cloudstackdeploymentconfigs.%s", v1alpha1.GroupVersion.Group)
```
Comment (maxdrib) [inline pkg/executables/kubectl.go:40]: guillermo suggests creating an initial PR with just the kubectl changes initially, then we can come back to the provider code

File: pkg/providers/cloudstack/cloudstack.go [hunk 3]
```diff
@@ -0,0 +1,1159 @@
...
+	GetSecret(ctx context.Context, secretObjectName string, opts ...executables.KubectlOpt) (*corev1.Secret, error)
+	UpdateAnnotation(ctx context.Context, resourceType, objectName string, annotations map[string]string, opts ...executables.KubectlOpt) error
+	SearchCloudStackMachineConfig(ctx context.Context, name string, kubeconfigFile string, namespace string) ([]*v1alpha1.CloudStackMachineConfig, error)
+	SearchCloudStackDatacenterConfig(ctx context.Context, name string, kubeconfigFile string, namespace string) ([]*v1alpha1.CloudStackDatacenterConfig, error)
+	DeleteEksaCloudStackDatacenterConfig(ctx context.Context, cloudstackDeploymentConfigName string, kubeconfigFile string, namespace string) error
+	DeleteEksaCloudStackMachineConfig(ctx context.Context, cloudstackMachineConfigName string, kubeconfigFile string, namespace string) error
+}
+
+type ClusterResourceSetManager interface {
+	ForceUpdate(ctx context.Context, name, namespace string, managementCluster, workloadCluster *types.Cluster) error
```
[the remark this answers is not in the document]
Comment (maxdrib) [reply pkg/providers/cloudstack/cloudstack.go:159]: Not needed as far as I can tell - removing

File: pkg/providers/cloudstack/cloudstack.go [hunk 4]
```diff
@@ -0,0 +1,1159 @@
...
+			}
+		}
+		etcdUser.SshAuthorizedKeys[0] = p.etcdSshAuthKey
+	}
+	controlPlaneUser.SshAuthorizedKeys[0] = p.controlPlaneSshAuthKey
+	workerUser.SshAuthorizedKeys[0] = p.workerSshAuthKey
+	return nil
+}
+
+func (p *cloudstackProvider) setupSSHAuthKeysForUpgrade() error {
```
[the remark this answers is not in the document]
Comment (maxdrib) [reply pkg/providers/cloudstack/cloudstack.go:331]: done

**Comment by chrisdoherty4** [comment]
Hey, I noticed this is a > 10,000 LOC PR with unit tests still to be written. I appreciate signiciant amounts of it are yaml, but I wonder if we could break it down into smaller more manageable PRs that can be reviewed more easily?

**Comment by maxdrib** [comment]
@chrisdoherty4 Thanks for the suggestion. I've actually been breaking this PR down gradually and submitting smaller PR's (e.g. https://github.com/aws/eks-anywhere/pull/1219 https://github.com/aws/eks-anywhere/pull/1237) and getting some feedback from the team that way, this is just for reference

**Comment by maxdrib** [comment]
Outdated and no longer needed. Closing

**Pull request closed without merging**